src/backend/PortabilityService.Functions/Report.cs

```diff
@@ -3,100 +3,115 @@
-            if (new Random().Next(10) < 4)
+            var (analyzeRequest, analyzeResult) = await GetSubmissionDataAsync(submissionId, storage, log);
+            if (analyzeRequest == null)
             {
-                // TODO this should return 202 only when the analyze request is known to have been received,
-                // 404 otherwise (e.g. when the client tries to get the report before its analyze request
-                // has propagated)
-                var response = req.CreateResponse(HttpStatusCode.Accepted);
-                response.Headers.RetryAfter = new RetryConditionHeaderValue(TimeSpan.FromSeconds(2d));
+                // the request may have been received but not yet stored => the client should poll again
+                var res = req.CreateResponse(HttpStatusCode.NotFound);
+                res.Headers.RetryAfter = new RetryConditionHeaderValue(TimeSpan.FromSeconds(4d));
+            }
+            if (analyzeResult == null)
+            {
+                // AnalyzeRequest received but AnalyzeResult not available => client should continue polling
+                var res = req.CreateResponse(HttpStatusCode.Accepted);
+                res.Headers.RetryAfter = new RetryConditionHeaderValue(TimeSpan.FromSeconds(2d));
 
-                return response;
+                return res;
             }
 
-            switch (req.Headers.Accept.ToString())
+            var response = req.CreateResponse();
+            try
+            {
+                response.Content = GetReportContent(req.Headers.Accept.SingleOrDefault(), analyzeRequest, analyzeResult);
+                response.StatusCode = HttpStatusCode.OK;
+            }
+            catch (UnsupportedMediaTypeException ex)
             {
-                case MediaType.Json:
-                    return JsonReport(req);
-                case MediaType.Excel:
-                    return ExcelReport(req);
-                default:
-                    return req.CreateResponse(HttpStatusCode.UnsupportedMediaType);
+                log.LogError("unknown format {MediaType} requested", ex.MediaType);
+                response.StatusCode = HttpStatusCode.UnsupportedMediaType;
             }
+
+            return response;
         }
 
-        public static bool ValidAccessKey(HttpRequestMessage request)
+        public static async Task<(AnalyzeRequest, AnalyzeResult)> GetSubmissionDataAsync(string submissionId, IStorage storage, ILogger log)
         {
-            // TODO generate and persist a new unique key in Analyze, validate it here
-            var authHeader = request.Headers.Authorization;
-            if (authHeader == null || !authHeader.Scheme.Equals("Bearer", StringComparison.Ordinal))
+            AnalyzeRequest analyzeRequest = null;
+            AnalyzeResult analyzeResult = null;
+            try
             {
-                return false;
+                analyzeRequest = await storage.RetrieveRequestAsync(submissionId);
```

Comment:
Can you move the variable declarations and definitions together? You can include the `return (analyzeRequest, analyzeResult)` in the try statement

Reply:
I don't see a better way. If the declarations and return are in the `try` block, the outer scope must return something ugly like `(null, null)` and lose the capability of returning an instance of `analyzeRequest` when assigning `analyzeResult` throws, which would alter the Function's response. But maybe I'm just not seeing it--I'm grateful for any suggestions here.

Reply:
ah - I thought it was a finally or a catch-rethrow block. I'd still prefer the `(null,null)` as it makes it easier to reason about what is happening